### PR TITLE
Give the SDK the trace and span entities

### PR DESCRIPTION
**Pull request by harry-rhesis**
Traces were the only substantial platform entity with no SDK class. `rhesis.sdk.telemetry` is the instrumentation side and writes traces; nothing read them back. So a script could fetch a test result but not what the endpoint actually did to produce it, which is where a failure explains itself.

First of three. This one is the SDK entity. PR-2 gives the same reach to MCP, the skill and the Architect, which today have zero trace tools out of 56 and say so outright in `skills/rhesis/references/entity-model.md:51`. PR-3 is docs.

## A trace has two ids and they are not interchangeable

This is the thing most likely to produce silently wrong code, and it already caused one documented trap in the Architect prompt.

`trace_id` is the OpenTelemetry hex, and it is what `GET /telemetry/traces/{trace_id}` takes. The root span's row id is what annotations, comments, tasks and the platform's own `/traces/{id}` URL take. Neither the list response nor the detail response carries that row id at the top level: the only route to it is `root_spans[0].id`.

`Trace` hides the difference. `db_id` resolves the row id from the span tree, fetching the detail once if the trace came from a listing, and `annotate()` and `get_annotations()` use it themselves.

```python
trace = Traces.for_test_run(run.id)[0]
trace.annotate("fail", "Answered from the wrong document.")
trace.span("ai.llm.invoke").annotate("fail", "Ignored the retrieved context.")
```

`Span.annotate()` exists because `TraceAnnotationsTab.tsx:121` passes `entityId={selectedSpan.id}`, so the UI annotates whichever span is selected. Without it the SDK would be less capable than the UI, and callers would hand-build `Annotations.create("Trace", span.id, ...)`, which is the thing the #2765 ergonomics work removed.

A trace with no root span row id raises instead of returning None. Returning None would file the annotation against nothing at all.

## Spans: the way back from a row id

The row id is the only id anything hands you. An annotation carries it as `context.trace_db_id`, and so do comments, tasks and trace links, none of which carry the hex. Nothing could turn it back into a trace.

```python
trace = Spans.trace_for(annotation.context.trace_db_id)
span = Spans.pull(row_id)   # resolves the trace, picks the span out of its tree
```

`pull` walks the tree rather than calling a route, because there is no route that returns a span on its own. `all()` and `first()` raise, each naming `Traces.query(root_spans_only=False)` as the way to list spans as rows.

`Traces.providers()` is there so the `provider` filter's valid values are fetched rather than guessed.

## Named filters, not OData

`Traces.query()` spells out all twenty filters. The route has no `$filter`, and adding one was considered and rejected: the list selects over `models.Trace` with correlated subqueries and a window function for the total, and does not go through `QueryBuilder`, which is what `with_odata_filter` plugs into.

Coverage decided it rather than risk. About eleven of the twenty filters are real columns OData could reach, while `provider`, `trace_source`, `trace_type`, `trace_metrics_status`, `has_errors` and `root_spans_only` come from JSONB attributes or subqueries. A filter syntax that handles `status_code` but not `provider` is worse than named params that handle both, because the failure is a filter that looks right and quietly does not apply. The frontend depends on the named params too, so `$filter` would be additive and we would carry two syntaxes regardless.

Only arguments the caller passed are sent, so the route's own defaults stay authoritative instead of being shadowed by a copy here that then drifts. The drop test is `is not None`, not truthiness: `root_spans_only=False` and `duration_min_ms=0` are real filters, and dropping them as falsy would quietly widen the query. That is the same bug as `turn=0` in #2765, and both cases are tested.

`limit` is the most traces to return in total, not the page size the route calls by that name. Omit it and every page is read.

## The inherited collection methods all assumed something this route does not do

They assume a bare list and a row id. This route returns `{traces, total, limit, offset}` and is addressed by a hex plus a required project. Each is overridden rather than left to fail obscurely:

- `all()` refuses an OData filter and points at `query()`, rather than sending one that would be ignored.
- `first()` asks for one trace instead of indexing into a dict.
- `pull()` takes the hex and a project, falling back to `RHESIS_PROJECT_ID`, and says so when it has neither. This is the one entity whose detail route needs a project rather than taking it from ambient scope.
- `exists()` reports a trace that was never ingested as False instead of raising on a 422.
- `for_test_run()` uses the run's own `/test_runs/{id}/traces`, because that route resolves the project from the run. Filtering the trace list by `test_run_id` needs the caller scoped to the right project first, and answers with an empty list rather than an error when they are not.

`push()` and `delete()` raise, naming `rhesis.sdk.telemetry`. The only write to a trace is ingestion.

## Two response shapes, one class

The list route returns the relations flat as ids; the detail route returns them nested as objects and no ids at all. A validator normalises on the way in, so `trace.test_run_id` is populated whichever call produced the trace, rather than being None on exactly the half of cases a caller did not anticipate. Same for the root operation and status, which a listing puts at the top level and the detail route only per span.

There is deliberately no `endpoint` field on `Trace`: that name is the API route pointer on every `BaseEntity`, and shadowing it would break `_pull` and `_create`. The endpoint is `endpoint_id` / `endpoint_name` plus `get_endpoint()`.

## Departures from the plan

`get_comments()`, `add_tag()` and `remove_tag()` were in the plan and are not here. No SDK entity has them, so adding them to `Trace` alone would make traces the only taggable entity. Comments and tags across every entity are their own piece of work.

The planned signature wrote `root_spans_only: bool = True` and `sort_order = "desc"`, which contradicts "only what the caller passed is sent" in the same paragraph. Both are `Optional[...] = None`, with the route's default named in the docstring.

## Tests

76 new tests in `tests/sdk/entities/test_trace.py`; the full SDK suite is at 3157 passed.

The edge cases are the point, so I checked the tests actually bite by mutating the source and confirming the intended test failed each time: truthiness instead of `is not None` (caught by the `False` and `0` filter tests), dropping the bare-string `provider` guard so `"openai"` becomes six letters (caught), dropping the flag that stops a trace with no spans being refetched forever (caught), and annotating by the hex instead of the row id (caught).

One caveat worth stating: removing the short-page break in the paging loop makes the empty-response test hang rather than fail, since an empty page leaves the offset where it was. The guard is load-bearing and the test proves it, but a regression there would stall CI instead of reporting.

## Integration tests

27 more in `tests/sdk/integration/test_trace.py`, against the live backend. The unit tests run on fixtures written by reading the backend schemas, which by construction cannot catch the SDK naming a field the route does not return. Traces are where that matters most: the only entity whose list and detail routes return different shapes, and the only one addressed by two different ids.

Traces are ingested through the OTEL route, which is how a trace comes into being, then read back through the SDK. What this catches that mocks cannot:

- The row id a listing does not carry really is `root_spans[0].id`. Mutating `annotate` to use the hex instead gives a real 404, rather than a quietly misfiled annotation.
- `root_spans_only=False` and `duration_min_ms=0` reach the route as real filters. Restoring the falsy guard turns the first back into one row instead of two, against the real query.
- Annotating a trace flips its `verdict`, so the override semantics hold end to end and not only in the request body.
- Every filter is checked against a trace it should match and one it should not, since a wrong parameter name is ignored server-side rather than refused.

Two things the harness needed, both of which say something about the routes. Span names are validated against the `ai.<domain>.<action>` and `function.<name>` convention, so the per-test unique name uses the `function.` form. The test organization is created with raw SQL and never went through seeding, so it has no statuses at all; a module fixture creates the `EntityType` lookup and the Pass/Fail rows a verdict points at, which is also why annotations have had no integration coverage until now.

Traces are never truncated between tests, so every assertion is scoped by a name or trace id unique to its test rather than by a total. Full integration suite: 173 passed.

### CODE REVIEW

File: sdk/src/rhesis/sdk/entities/trace.py [hunk 1]
```diff
@@ -0,0 +1,714 @@
...
+            return
+        self._merge(self._fetch_detail())
+
+    def _merge(self, payload: Dict[str, Any]) -> None:
+        fresh = Trace.model_validate(payload)
+        for name in type(self).model_fields:
+            value = getattr(fresh, name)
+            # A listing carries fields the detail route does not, so only what
+            # came back is applied; the rest of this trace stays as it was.
+            if value is not None and value != []:
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply sdk/src/rhesis/sdk/entities/trace.py]: Valid, and fixed in 7bb908440.

You are right that the rule was wrong, though the reachability is narrower than it looks: the detail route 404s when a trace has no spans, and `build_span_tree` promotes an orphan whose parent is missing to a root, so a 200 always carries at least one root span. An empty `root_spans` cannot come back today.

That is exactly why I did not want to special-case `root_spans`. It would have been a second guess layered on the first, and the rule would still have depended on a backend invariant nothing in the SDK states.

The real question is not what the value looks like but whether the response carried the field at all, which is what `model_fields_set` answers:

```python
for name in fresh.model_fields_set:
    setattr(self, name, getattr(fresh, name))
```

This keeps both halves honest. Absent stays distinct from empty, so the detail route merging in no longer blanks `models` and `providers` that only a listing reports, and an explicit empty `root_spans` is now applied rather than skipped. It also survives the `@model_validator(mode="before")`: the ids flattened out of the detail route's nested relations count as set, because the validator puts them in the dict being validated, so `endpoint_id` and `root_operation` still apply.

Two tests, one for each direction: a re-pull no longer keeps a span tree the server stopped reporting, and a re-pull still keeps what only a listing carries. I checked the first one bites by restoring the old rule and watching it fail.

File: sdk/src/rhesis/sdk/entities/trace.py [hunk 2]
```diff
@@ -0,0 +1,714 @@
...
+            response = (
+                client.send_request(
+                    endpoint=endpoint or cls.endpoint,
+                    method=Methods.GET,
+                    url_params=url_params,
+                    params={**params, "offset": offset, "limit": page_size},
+                )
+                or {}
+            )
+            page = response.get("traces") or []
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply sdk/src/rhesis/sdk/entities/trace.py:416]: Fail fast is intended, and I have added the note in 7bb908440 rather than the try/except.

Partial results would be the wrong trade here specifically. This collection already refuses to truncate: `_paged` reads every page rather than stopping at the first, because a list that is quietly short is the worst way to be wrong about traces, and the caller has no way to tell it happened. Skipping a malformed row is the same failure in a smaller package. You would get 99 traces back, no error, and no reason to suspect the hundredth existed.

It is also a narrow failure mode. Every field on `Trace` is optional with a default and unknown keys are ignored, so a row can only fail validation on a type mismatch, which means the route changed shape rather than one trace being odd. That is a bug to surface, not to route around, and if it happens it will happen to every row rather than one.

The docstring now says so:

> A row the model cannot read raises rather than being skipped. Dropping it would hand back a list that is quietly short, which is the same failure as stopping at the first page, and the caller would have no way to tell. Every field is optional, so this means the route changed shape.

File: sdk/src/rhesis/sdk/entities/trace.py [hunk 3]
```diff
@@ -0,0 +1,714 @@
...
+
+    @classmethod
+    def pull(cls, trace_id: str, project_id: Optional[str] = None) -> Trace:
+        """One trace with its full span tree, by its OpenTelemetry ``trace_id``.
+
+        Unlike every other entity, the route needs the project: pass
+        ``project_id`` or set ``RHESIS_PROJECT_ID``. Traces have no names, so
+        there is nothing to look one up by.
+        """
+        trace = Trace(trace_id=trace_id, project_id=project_id)
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply sdk/src/rhesis/sdk/entities/trace.py:538]: Intentional, and I want to push back on the suggested fix specifically: aliasing `id` to `trace_id` would re-introduce the exact bug this entity exists to prevent.

A trace has two ids and they are not interchangeable. `trace_id` is the OpenTelemetry hex, which only the detail route takes. The root span's row id is what annotations, comments, tasks and the platform's own `/traces/{id}` URL take. Generic tooling that reads `.id` is, almost by definition, about to use it to address the entity, and handing it the hex means it addresses nothing. The integration test in this PR demonstrates that concretely: mutating `annotate` to send `trace_id` instead of `db_id` is a real 404 from the live backend.

Aliasing `id` to `db_id` would be correct but worse in a different way, since resolving it can require a network call, and an attribute that silently issues a request is not something generic tooling should walk into.

So the honest shape is the one that is there: `trace_id` and `db_id` named for what they are, with the docstring stating the distinction up front, and nothing called `id` to be grabbed by accident.

On the convention question, nothing in the SDK breaks. The only generic `.id` access in the base classes is `BaseEntity.push` at line 153, and `push`, `delete` and the instance `pull` are all overridden on `Trace` to raise or to address by `trace_id`; `BaseCollection.pull`, `first`, `all` and `exists` are overridden for the same reason. I checked rather than assumed. A caller who writes `Traces.pull(id=...)` gets a `TypeError` naming the argument, which is the loud failure, not a request against the wrong id.

File: sdk/src/rhesis/sdk/entities/trace.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply sdk/src/rhesis/sdk/entities/trace.py]: All three points were well-spotted initial observations. The `model_fields_set` approach in `_merge()` handles the absent-vs-empty distinction correctly, the fail-fast validation is consistent with the rest of the SDK, and keeping `trace_id`/`db_id` explicit avoids ambiguity. Thanks for the thorough self-investigation on each thread.
